File: app/src/main/java/com/mdeiml/richard/SavedGamesAdapter.java
package com.mdeiml.richard;

import android.graphics.Typeface;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.content.Context;
import android.widget.CursorAdapter;
import android.view.View;
import android.view.ViewGroup;
import android.view.LayoutInflater;
import android.widget.TextView;
import android.widget.TableRow;
import android.util.TypedValue;
import java.util.Locale;

public class SavedGamesAdapter extends CursorAdapter {
    
    private final int idIndex;
    private final int player1Index;
    private final int player2Index;
    private final int winnerIndex;
    private SavedMatchesDbHelper dbHelper;
    private int runningHeaderIndex;
    private int overHeaderIndex;

    public SavedGamesAdapter(Context context, SavedMatchesDbHelper dbHelper) {
        super(context, createCursor(dbHelper.getReadableDatabase(), context), false);
        this.idIndex = getCursor().getColumnIndexOrThrow("_id");
        this.player1Index = getCursor().getColumnIndexOrThrow("match_player1");
        this.player2Index = getCursor().getColumnIndexOrThrow("match_player2");
        this.winnerIndex = getCursor().getColumnIndexOrThrow("match_winner");
        this.dbHelper = dbHelper;
        this.runningHeaderIndex = -1;
        this.overHeaderIndex = -1;
    }

    private static Cursor createCursor(SQLiteDatabase db, Context context) {
        return db.rawQuery(context.getString(R.string.sql_saved_matches), new String[0]);
    }

    public int getViewTypeCount() {
        return 2;
    }

    public int getItemViewType(int position) {
        Cursor cursor = getCursor();
        cursor.moveToPosition(position);
        int winner = cursor.getInt(winnerIndex);
        if (position == runningHeaderIndex || position == overHeaderIndex) {
            return 0;
        } else if (winner == -1) {
            runningHeaderIndex = position;
            return 0;
        } else if (winner == 1) {
            overHeaderIndex = position;
            return 0;
        } else {
            return 1;
        }
    }

    public void bindView(View view, Context context, Cursor cursor) {
        int winner = cursor.getInt(winnerIndex);
        if (winner == -1) {
            ((TextView) view).setText(context.getString(R.string.game_running_heading));
        } else if (winner == 1) {
            ((TextView) view).setText(context.getString(R.string.game_over_heading));
        } else {
            final long matchId = cursor.getLong(idIndex);
            final int position = cursor.getPosition();
            TextView sgPlayer1 = (TextView)view.findViewById(R.id.saved_game_player1);
            TextView sgPlayer2 = (TextView)view.findViewById(R.id.saved_game_player2);
            TextView sgPoints1 = view.findViewById(R.id.saved_game_points1);
            TextView sgPoints2 = view.findViewById(R.id.saved_game_points2);
            TableRow sgSets1 = (TableRow)view.findViewById(R.id.saved_game_sets1);
            TableRow sgSets2 = (TableRow)view.findViewById(R.id.saved_game_sets2);

            Match match = dbHelper.loadMatch(matchId);
            sgPlayer1.setText(match.player1);
            sgPlayer2.setText(match.player2);

            int numSets = match.getCurrentSetNr()+1;
            byte[][] games = match.getGames();
            for (int i = 0; i < numSets; i++) {
                TextView tv0 = (TextView) sgSets1.getChildAt(i + 1);
                tv0.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20);
                tv0.setText(String.format(Locale.getDefault(), "%d", games[i][0]));
                tv0.setLayoutParams(new TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 1));
                tv0.setVisibility(View.VISIBLE);
                
                TextView tv1 = (TextView) sgSets2.getChildAt(i + 1);
                tv1.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20);
                tv1.setText(String.format(Locale.getDefault(), "%d", games[i][1]));
                tv1.setLayoutParams(new TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 1));
                tv1.setVisibility(View.VISIBLE);
            }

            for (int i = numSets; i < 3; i++) {
                sgSets1.getChildAt(i + 1).setVisibility(View.INVISIBLE);
                sgSets2.getChildAt(i + 1).setVisibility(View.INVISIBLE);
            }

            String[] scores = match.getCurrentSet().getCurrentGame().stringScores();
            if(scores[0] != "0" || scores[1] != "0") {
                sgPoints1.setText(scores[0]);
                sgPoints2.setText(scores[1]);
            }
        }
    }
    
    public View newView(Context context, Cursor cursor, ViewGroup parent) {
        int winner = cursor.getInt(winnerIndex);
        if (winner == -1 || winner == 1) {
            TextView heading = new TextView(context);
            heading.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20);
            heading.setTextColor(0xff707070);
            heading.setTypeface(null, Typeface.BOLD);
            return heading;
        } else {
            LayoutInflater inflater = LayoutInflater.from(context);
            return inflater.inflate(R.layout.saved_game, parent, false);
        }
    }
    
    public void updateCursor(Context context) {
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        Cursor cursor = createCursor(db, context);
        changeCursor(cursor);
        runningHeaderIndex = -1;
        overHeaderIndex = -1;
    }

    public boolean isHeader(int position) {
        return position != -1 && (position == runningHeaderIndex || position == overHeaderIndex);
    }
}

File: app/src/main/kotlin/com/mdeiml/richard/SavedGamesAdapter.kt
package com.mdeiml.richard

import android.graphics.Typeface
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.content.Context
import android.widget.CursorAdapter
import android.view.View
import android.view.ViewGroup
import android.view.LayoutInflater
import android.widget.TextView
import android.widget.TableRow
import android.util.TypedValue

public class SavedGamesAdapter(
        context: Context,
        dbHelper: SavedMatchesDbHelper
) : CursorAdapter(
        context,
        SavedGamesAdapter.createCursor(dbHelper.getReadableDatabase(), context),
        false
) {
    
    private val idIndex = cursor.getColumnIndexOrThrow("_id")
    private val player1Index = cursor.getColumnIndexOrThrow("match_player1")
    private val player2Index = cursor.getColumnIndexOrThrow("match_player2")
    private val winnerIndex = cursor.getColumnIndexOrThrow("match_winner")
    private val dbHelper = dbHelper
    private var runningHeaderIndex: Int? = null
    private var overHeaderIndex: Int? = null

    companion object {
        fun createCursor(db: SQLiteDatabase, context: Context): Cursor {
            return db.rawQuery(context.getString(R.string.sql_saved_matches), arrayOf())
        }
    }

    override fun getViewTypeCount() = 2

    override fun getItemViewType(position: Int): Int {
        if (position == runningHeaderIndex || position == overHeaderIndex) {
            return 0
        }

        cursor.moveToPosition(position)
        val winner = cursor.getInt(winnerIndex)
        when (winner) {
            -1 -> {
                runningHeaderIndex = position
                return 0
            }
            1 -> {
                overHeaderIndex = position
                return 0
            }
            else -> return 1
        }
    }

    override fun isEnabled(position: Int) = getItemViewType(position) != 0

    override fun bindView(view: View, context: Context, cursor: Cursor) {
        val maxSets = 3 // TODO
        val locale = context.resources.configuration.locales[0]

        val winner = cursor.getInt(winnerIndex)
        when (winner) {
            -1 -> (view as TextView).setText(context.getString(R.string.game_running_heading))
            1 -> (view as TextView).setText(context.getString(R.string.game_over_heading))
            else -> {
                val matchId = cursor.getLong(idIndex)

                val sgPlayer1 = view.findViewById(R.id.sgPlayer1) as TextView
                val sgPlayer2 = view.findViewById(R.id.sgPlayer2) as TextView
                val sgPoints1 = view.findViewById(R.id.sgPoints1) as TextView
                val sgPoints2 = view.findViewById(R.id.sgPoints2) as TextView
                val sgSets1 = view.findViewById(R.id.sgSets1) as ViewGroup
                val sgSets2 = view.findViewById(R.id.sgSets2) as ViewGroup

                val match = dbHelper.loadMatch(matchId)!!
                sgPlayer1.text = match.player1 ?: context.getString(R.string.player_a)
                sgPlayer2.text = match.player2 ?: context.getString(R.string.player_b)

                val numSets = match.currentSetNr + if (match.currentSet.tiebreak) 0 else 1
                val games = match.gameScore
                for (i in 0..numSets-1) {
                    val tv0 = sgSets1.getChildAt(i + 1) as TextView
                    tv0.text = String.format(locale, "%d", games[i].first)
                    tv0.setTextColor(context.resources.getColor(R.color.textContent))
                    tv0.visibility = View.VISIBLE
                    
                    val tv1 = sgSets2.getChildAt(i + 1) as TextView
                    tv1.text = String.format(locale, "%d", games[i].second)
                    tv0.setTextColor(context.getResources().getColor(R.color.textContent))
                    tv1.visibility = View.VISIBLE
                }

                for (i in numSets..maxSets-1) {
                    sgSets1.getChildAt(i + 1).visibility = View.INVISIBLE
                    sgSets2.getChildAt(i + 1).visibility = View.INVISIBLE
                }

                val (s1, s2) = match.currentSet.currentGame.stringScores
                if(s1 != "0" || s2 != "0") {
                    sgPoints1.text = s1
                    sgPoints2.text = s2
                }
            }
        }
    }
    
    override fun newView(context: Context, cursor: Cursor, parent: ViewGroup): View {
        val winner = cursor.getInt(winnerIndex)
        when (winner) {
            -1, 1 -> {
                val heading = TextView(context)
                heading.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20.0f)
                heading.setTextColor(0xff707070.toInt())
                heading.setTypeface(null, Typeface.BOLD)
                return heading
            }
            else -> {
                val maxSets = 3 // TODO
                val inflater = LayoutInflater.from(context)
                val view = inflater.inflate(R.layout.saved_game, parent, false)

                val sgSets1 = view.findViewById(R.id.sgSets1) as ViewGroup
                val sgSets2 = view.findViewById(R.id.sgSets2) as ViewGroup
                for (i in 0..maxSets-1) {
                    val tv0 = TextView(context)
                    tv0.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20.0f)
                    tv0.layoutParams = TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 1.0f)
                    sgSets1.addView(tv0, i + 1)
                    
                    val tv1 = TextView(context)
                    tv1.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20.0f)
                    tv1.layoutParams = TableRow.LayoutParams(0, TableRow.LayoutParams.WRAP_CONTENT, 1.0f)
                    sgSets2.addView(tv1, i + 1)
                }
                return view
            }
        }
    }
    
    fun updateCursor(context: Context) {
        val db = dbHelper.getReadableDatabase()
        val cursor = createCursor(db, context)
        changeCursor(cursor)
        runningHeaderIndex = null
        overHeaderIndex = null
    }
}
